### PR TITLE
Feature/waves

### PR DESCRIPTION
**Pull request by Mikefluff**
Initial waves support

### CODE REVIEW

File: packages/login/network/WavesEngine.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+
+  createTransaction (type, params) {
+
+    const data = this.describeTransaction(type, params)
+
+    const transferTransaction = new this._Transactions.TransferTransaction(data);
+
+    const api = transferTransaction.prepareForAPI(keys.privateKey).then((preparedData) => {
+      return preparedData
```
Comment (ipavlenko) [inline packages/login/network/WavesEngine.js]: Odd code

File: packages/login/network/WavesNode.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ * Copyright 2017–2018, LaborX PTY
+ * Licensed under the AGPL Version 3 license.
+ */
+
+import axios from 'axios'
+import WavesMiddlewareNode from './WavesMiddlewareNode'
+import * as WavesApi from '@waves/waves-api'
+
+export const WAVES_MAINNET_NODE = new WavesMiddlewareNode({
+  assets: ['MINUTES'],
```
Comment (ipavlenko) [inline packages/login/network/WavesNode.js]: Do we have minutes on WAVES?

File: packages/login/network/WavesProvider.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      time: new Date().getTime(),
+      tx,
+    })
+  }
+
+  async onBalance (balance: WavesBalance) {
+    this.emit('balance', {
+      account: this.getAddress(),
+      time: new Date().getTime(),
+      balance, // structured for NEM
```
Comment (ipavlenko) [inline packages/login/network/WavesProvider.js]: Wrong comment

File: src/dao/WavesDAO.js [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+      this.watchTransfer(),
+      this.watchBalance(),
+    ])
+  }
+
+  async watchTransfer () {
+    this._wavesProvider.addListener(EVENT_TX, async ({ tx }) => {
+      if (tx.unconfirmed) {
+        if (!this._mosaic) {
+          if (!tx.mosaics) {
```
Comment (ipavlenko) [inline src/dao/WavesDAO.js]: Mosaics or Assets?

Mosaics for NEM,
Assets for Waves

It seems the code related to Assets is not fully ported

File: src/dao/WavesDAO.js [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+            this.emit(EVENT_NEW_TRANSFER, this._createXemTxModel(tx))
+          }
+        } else {
+          if (tx.mosaics && (this._namespace in tx.mosaics)) {
+            this.emit(EVENT_NEW_TRANSFER, this._createMosaicTxModel(tx))
```
Comment (ipavlenko) [inline src/dao/WavesDAO.js]: Wrong code, _createMosaicTxModel relates to NEM

File: src/dao/WavesDAO.js [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+      isFetched: true,
+      blockchain: BLOCKCHAIN_WAVES,
+    })
+
+    return token
+  }
+}
+
+//TODO WHY WE NEED SYMBOL AND ASSET DESCRIPTION IF SYMBOL IS ENOUGH
+function readBalanceValue (symbol, balance, asset = null) {
```
Comment (ipavlenko) [inline src/dao/WavesDAO.js]: Mosaic description required to sign NEM Mosaic transfers + Mosaic definition contains info about decimals places in Mosaic amounts.
Do not sure if we need full asset description for Waves.

File: src/redux/mainWallet/actions.js [hunk 7]
```diff
@@ -7,7 +7,7 @@ import {
   bccProvider,
   BLOCKCHAIN_BITCOIN,
   BLOCKCHAIN_BITCOIN_CASH,
-  BLOCKCHAIN_BITCOIN_GOLD,
+  BLOCKCHAIN_BITCOIN_GOLD,
```
Comment (ipavlenko) [inline src/redux/mainWallet/actions.js]: linter issue

File: src/redux/mainWallet/actions.js [hunk 8]
```diff
@@ -17,6 +17,7 @@ import { ethereumProvider } from '@chronobank/login/network/EthereumProvider'
 import { change, formValueSelector } from 'redux-form/immutable'
 import { history } from 'redux/configureStore'
 import { nemProvider } from '@chronobank/login/network/NemProvider'
+import { wavesProvider} from '@chronobank/login/network/WavesProvider'
```
Comment (ipavlenko) [inline src/redux/mainWallet/actions.js]: linter issue

File: packages/login/network/EthereumMiddlewareNode.js [hunk 9]
```diff
@@ -45,19 +46,25 @@ export default class EthereumMiddlewareNode extends AbstractNode {
       })
 
     } catch (e) {
+      console.log('Address subscription error')
```
Comment (ozalexo) [inline packages/login/network/EthereumMiddlewareNode.js]: trace and log at the same time? Maybe console.log should be removed?

File: packages/login/network/WavesEngine.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+      // ID of a token, or WAVES
+      assetId: asset,
+
+      // The real amount is the given number divided by 10^(precision of the token)
+      amount: amount,
+
+      // The same rules for these two fields
+      feeAssetId: 'WAVES',
+      fee: 100000,
```
Comment (ozalexo) [inline packages/login/network/WavesEngine.js]: "Magic number". Additional comment would be great.

File: packages/login/network/WavesNode.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+  api: axios.create({
+    baseURL: 'https://middleware-waves-mainnet-rest.chronobank.io',
+    timeout: 30000,
+  }),
+  socket: {
+    baseURL: 'https://rabbitmq-webstomp.chronobank.io/stomp',
+    user: 'rabbitmq_user',
+    password: '38309100024',
```
Comment (ozalexo) [inline packages/login/network/WavesNode.js]: Hmmm... Password in source code?

File: packages/login/network/WavesWallet.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+export default class WavesWallet {
+  constructor (seed, keyPair, network) {
+    this._seed = seed
+    this._keyPair = keyPair
+    this._network = network
+  }
+
+  getAddress () {
+    console.log('Waves address:')
+    console.log(this._seed.address)
```
Comment (wAISw) [inline packages/login/network/WavesWallet.js]: Remove logs

File: packages/login/network/mnemonicProvider.js [hunk 13]
```diff
@@ -57,19 +62,23 @@ class MnemonicProvider {
       const btgNetwork = network.bitcoinGold && bitcoin.networks[network.bitcoinGold]
       const ltcNetwork = network.litecoin && bitcoin.networks[network.litecoin]
       const nemNetwork = network.nem && nemSdk.model.network.data[network.nem]
+      const wavesNetwork = network.waves && WavesApi[network.waves]
+      console.log('Waves network is:')
+      console.log(wavesNetwork)
```
Comment (wAISw) [inline packages/login/network/mnemonicProvider.js]: Remove logs

File: src/redux/mainWallet/actions.js [hunk 14]
```diff
@@ -228,6 +231,11 @@ export const fetchTokenBalance = (token: TokenModel) => async (dispatch, getStat
   const tokenDAO = tokenService.getDAO(token.id())
   const { account } = getState().get(DUCK_SESSION)
   const balance = await tokenDAO.getAccountBalance(token.blockchain() === BLOCKCHAIN_ETHEREUM ? account : null)
+  if(token.blockchain() === BLOCKCHAIN_WAVES) {
```
Comment (ozalexo) [inline src/redux/mainWallet/actions.js]: It seems like definitely debug console.logs...

File: src/components/common/ui/Button/Button.jsx [hunk 15]
```diff
@@ -75,7 +75,7 @@ export default class Button extends PureComponent {
           disabled={this.props.disabled}
           styleName={classnames('button', buttonType)}
           type={this.props.type}
-          onTouchTap={this.handleTouchTap}
+          onClick={this.handleTouchTap}
```
Comment (wAISw) [inline src/components/common/ui/Button/Button.jsx]: Please, change handleTouchTap to handleOnClick

**Pull request merged**